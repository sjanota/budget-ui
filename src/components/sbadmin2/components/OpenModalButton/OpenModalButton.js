import React, { useState } from 'react';
import PropTypes from 'prop-types';

export default function OpenModalButton({
  renderButton,
  renderModal,
  onClickRef,
}) {
  const [show, setShow] = useState(false);
  const onHide = () => setShow(false);
  const onClick = () => setShow(true);
  if (onClickRef) {
    onClickRef.current = onClick;
  }
  return (
    <>
      {renderButton({ onClick })}
      {renderModal({ onHide, show })}
    </>
  );
}

OpenModalButton.propTypes = {
  renderButton: PropTypes.elementType.isRequired,
  renderModal: PropTypes.elementType.isRequired,
  onClickRef: PropTypes.shape({ current: PropTypes.any }),
};
