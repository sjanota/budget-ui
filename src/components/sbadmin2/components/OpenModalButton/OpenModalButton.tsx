import React, { ComponentType, MouseEvent, useState } from 'react';
import { Modal } from 'react-bootstrap';

interface ButtonProps {
  onClick(event: MouseEvent<HTMLButtonElement>): void;
}

interface ModalContentProps {
  onHide(): void;
}

interface Props {
  button: ComponentType<ButtonProps>;
  modalContent: ComponentType<ModalContentProps>;
}

export default function OpenModalButton({
  button: Button,
  modalContent: ModalContent,
  ...props
}: Props) {
  const [show, setShow] = useState<boolean>(false);
  const onHide = () => setShow(false);
  const onClick = () => setShow(true);

  return (
    <span>
      <Button onClick={onClick} />
      <Modal show={show} onHide={onHide} {...props}>
        <ModalContent onHide={onHide} />
      </Modal>
    </span>
  );
}
