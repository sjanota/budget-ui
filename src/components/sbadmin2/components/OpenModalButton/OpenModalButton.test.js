import renderer from 'react-test-renderer';
import OpenModalButton from './OpenModalButton';
import React from 'react';
import { mount } from 'enzyme';
import { actAndUpdate } from '../../../../testing';

it('OpenModalButton renders with minimal props', () => {
  const component = renderer.create(
    <OpenModalButton
      renderButton={props => <button {...props} />}
      renderModal={() => <div />}
    />
  );
  expect(component).toMatchSnapshot();
});

describe('OpenModalButton', () => {
  const renderButton = props => <button {...props} />;

  const Modal = () => <div />;
  const renderModal = props => <Modal {...props} />;
  const onClickRef = { current: null };

  const component = mount(
    <OpenModalButton
      renderButton={renderButton}
      renderModal={renderModal}
      onClickRef={onClickRef}
    />
  );

  it('modal is initialy hidden', () => {
    expect(component.find(Modal).prop('show')).toBe(false);
  });

  it('onClickRef points to button onClick', () => {
    expect(onClickRef.current).toBe(component.find('button').prop('onClick'));
  });

  it('click on rendered button opens modal', async () => {
    await actAndUpdate(component, () => {
      component.find('button').simulate('click');
    });

    expect(component.find(Modal).prop('show')).toBe(true);
  });

  it('modal onHide closes modal', async () => {
    await actAndUpdate(component, () => {
      component.find(Modal).prop('onHide')();
    });
    expect(component.find(Modal).prop('show')).toBe(false);
  });
});
