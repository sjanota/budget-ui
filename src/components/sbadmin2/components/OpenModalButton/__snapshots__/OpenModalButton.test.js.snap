// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OpenModalButton renders with minimal props 1`] = `
Array [
  <button
    onClick={[Function]}
  />,
  <div />,
]
`;
